teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md

```diff
@@ -0,0 +1,542 @@
+
+Generating the objects that satisfy these schemas, from human readable provenance metadata can be done by writing provenance data using Steps in Tasks into these schema structures in TaskRuns and populate them PipelineRuns' Results, and there is also a discussion around adding a new field `provenance` in PipelineRun and TaskRun, which only allows [Trusted Resources](https://github.com/tektoncd/community/blob/2413ad70a742a6e9103c531e1c5788b2b392a7eb/teps/0091-verified-remote-resources.md#requirements) to generate provenance info in this field. Chains should be able to accomodate what is currently capable in TaskRun and PipelineRun results, and other potential new fields specifically for provenances.
+
+### Notes and Caveats
+
+When these well defined structures grow as richer provenance are pursued, the size of the structure can eventually grow beyond the limit of the container termination message. [TEP 0086](https://github.com/tektoncd/community/pull/521) is addressing this issue.
+
+## Design Details
+
+### Signable Artifacts
+
+All signable artifacts should be provided in a structure that is well defined in the proposal. If any field is missed, the signable artifacts will be skipped and an error will be thrown. The users need to follow the naming pattern in order the result objects can be captured.
+
+As described above, users of Tekton Chains need to have two results, `ARTIFACT_INPUTS` and `ARTIFACT_OUTPUTS`, that are arrays of strings which stores the signable artifacts' result names:
+
+```yaml
+results:
+  - name: ARTIFACT_INPUTS
+    value: ["ARTIFACT-INPUT-NAME-1", "ARTIFACT-INPUT-NAME-1"]
+  - name: ARTIFACT_OUTPUTS
+    value: ["ARTIFACT-OUTPUT-NAME-1"]
+```
+
+And the signable target results need to follow the below schema:
+
+```yaml
+results:
+  - name: ARTIFACT-NAME
+    type: object
+    description: |
+      * uri: resource uri of the artifact. It can uniquely identify the artifact.
+      * digest: revision digest in form algorithm:digest.
+    properties:
+      uri:
+        type: string
+      digest:
+        type: string
+```
+
+In this way, every time when Tekton Chains tries to support a new type of artifacts, users can form in-toto provenance from this schema format, without needing to change Tekton Chains much. In general, Tekton Chains itself doesn't want to distinguish the types of artifacts from attestation generation point of view. The namings of these artifact results would be much more flexible as well.
+
+### DigestSets
+
+The in-toto provenance supports digest in terms of [DigestSets](https://github.com/in-toto/attestation/blob/main/spec/field_types.md#field-type-definitions) for an artifact, which means users can provide different digests for different digest SHAs. This would be hard to support in Chains, as Results doesn't support nested objects yet. In the proposal, only one pair of SHA and digest can be provided. `DigestSets` can be supported when nested objects are supported in `Results`.
```

Comment:
```suggestion
The in-toto provenance supports digest in terms of [DigestSets](https://github.com/in-toto/attestation/blob/main/spec/field_types.md#field-type-definitions) for an artifact, which means users can provide different digests for different digest SHAs. Since Pipeline Results cannot support nested objects yet, Chains is unable to support them at the moment. In the proposal, only one pair of SHA and digest can be provided. `DigestSets` can be supported when nested objects are supported in `Results`.
```

Rephrased - it's not that this is hard for Chains to support, it's that we're limited by what Pipelines supports.